onnxruntime/core/providers/cuda/tensor/transpose_impl.cu

```diff
@@ -133,38 +140,58 @@ bool CanDoTranspose4DParallelizeMultipleElementsPerThreadInInnermostDim(const cu
 }
 
 Status Transpose4DParallelizeMultipleElementsPerThreadInInnermostDim(
-    cudaStream_t stream, size_t element_size,
+    const cudaDeviceProp& prop, cudaStream_t stream, size_t element_size,
     const TArray<int64_t>& input_shape, const TArray<int64_t>& input_strides,
     const void* input_data, const TArray<int64_t>& output_strides,
     void* output_data, int N) {
   unsigned int num_elements_per_thread = 4 * sizeof(int) / static_cast<unsigned int>(element_size);  // int4 is used in the kernel to access data.
-  dim3 block_size(static_cast<unsigned int>(input_shape[3] / num_elements_per_thread), static_cast<unsigned int>(input_shape[2]));
-  dim3 grid_size(static_cast<unsigned int>(input_shape[1]), static_cast<unsigned int>(input_shape[0]));
+  int64_t num_block_ext;
+
+  if (input_shape[3] / num_elements_per_thread * 2 > prop.maxThreadsPerBlock) {
+    num_block_ext = input_shape[2];
+  } else {
+    int64_t num_elements_in_last_two_dimensions = input_shape[2] * input_shape[3];
+    num_block_ext = CeilDiv(num_elements_in_last_two_dimensions / num_elements_per_thread, prop.maxThreadsPerBlock);
```

Comment:
Here needs to be adjusted to make sure the blockDim.x * blockDim.y is multiples of warp size 32.